c2/cuda/groth16_srs.cuh

```diff
@@ -178,21 +178,28 @@ private:
                 } h, l, a, b_g1, b_g2;
             } data;
 
+            if (!ngpus()) {
+                throw groth16_error(-9, std::string("No CUDA devices available"));
+            }
+
             int srs_file = open(srs_path, O_RDONLY);
 
-            // TODO, replace asserts with custom exceptions
-            assert(srs_file >= 0);
+            if (srs_file < 0) {
+                throw groth16_error(-10, std::string("Failed to open SRS file"));
```

Comment:
Updated exception.hpp has a fmt_errno helper. For example in this case something like following could be more appropriate:
```
throw sppark_error{-errno, fmt_errno(errno, "open(\"%s\") failed: ", srs_path);
```

Reply:
I know that it was me who suggested this, but as I see it as actual code I wonder if it would be more appropriate to pass `errno`-s as is and negate the cudaError-s instead. Point would be that `errno` is something developers are more likely to be familiar with than the CUDA errors. And the message to the developers would be kind of "if the `code` in the RustError is positive, then it's likely your problem [and you might already know what it means], and if negative, it's something to report to us." Shall we exchange the signs?

Reply:
I like that idea! That means we would negate the sign [here](https://github.com/supranational/sppark/blob/main/util/exception.cuh#L19), too, right?